### ISSUE TITLE
IDE0059 incorrectly triggers when assigning to the indexer of a value type

### ISSUE THREAD

**Issue by tannergooding** [dotnet/roslyn#77258]
**Version Used**:  `Compiler version: '4.14.0-1.25105.1 (842cb1bc)'. Language version: latest (13.0).`

**Steps to Reproduce**:

1. Create a new C# Console Application in VS 2022 v17.14.0 Preview 1 (it may also occur earlier, I didn't test)
2. Update the code to the following:
```csharp
int[] a = new int[5];
MyStruct m = new MyStruct(a);
m[0] = 1; // IDE0059: Unnecessary assignment of a value to 'm'
Console.WriteLine(a[0]);

struct MyStruct(int[] a)
{
    private int[] array = a;

    public int this[int index]
    {
        get => array[index];
        set => array[index] = value;
    }
}
```
3. Observe that the line containing `m[0] = 1;` is flagged with ``

**Diagnostic Id**: `IDE0059: Unnecessary assignment of a value to '...'`

**Expected Behavior**:

The IDE warning should not trigger for method invocations or indexer usage, as these can trivially mutate indirect state. Such indirect side effects is the case in the example above where the contents of `a` are observably changed via the `m[0] = 1;`

**Comment by CyrusNajmabadi** [comment]
I believe this was fixed with: https://github.com/dotnet/roslyn/pull/77158

**Comment by CyrusNajmabadi** [comment]
@sharwell can you confirm?

**Comment by jbevain** [comment]
I'm seeing the diagnostic with Tanner's repro on `17.14.0 Preview 2.0 [35817.74.main]`.

My repro:

```csharp
unsafe class Program
{
    static void Main()
    {
        IO io = new(42);
        IOPtr ioPtr = new(&io);
        Settings(ioPtr);

        Console.WriteLine(io.Foo);
    }

    static void Settings(IOPtr ioPtr)
    {
        ioPtr.Foo = 12; // IDE0059: Unnecessary assignment of a value to 'ioPtr'
    }
}

struct IO(int foo)
{
    public int Foo = foo;
}

readonly unsafe struct IOPtr(IO* ptr)
{
    private readonly IO* _ptr = ptr;

    public ref int Foo => ref _ptr->Foo;
}
```
